File: android/src/main/java/com/wultra/android/powerauth/reactnative/PowerAuthModule.java
/*
 * Copyright 2020 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wultra.android.powerauth.reactnative;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Dynamic;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.wultra.android.powerauth.js.ActivityProvider;
import com.wultra.android.powerauth.js.PowerAuthJsModule;

import javax.annotation.Nonnull;

@SuppressWarnings("unused")
@ReactModule(name = "PowerAuth")
public class PowerAuthModule extends ReactContextBaseJavaModule {

    private final PowerAuthJsModule powerAuthJsModule;

    private final ReactApplicationContext context;
    private final ObjectRegister objectRegister;
    private final PowerAuthPasswordModule passwordModule;

    private final ActivityProvider activityProvider = () -> getCurrentActivity();

    public PowerAuthModule(ReactApplicationContext context, @NonNull ObjectRegister objectRegister, @NonNull PowerAuthPasswordModule passwordModule) {
        super(context);
        this.context = context;
        this.objectRegister = objectRegister;
        this.passwordModule = passwordModule;
        this.powerAuthJsModule = new PowerAuthJsModule(context, activityProvider, objectRegister.getObjectRegisterJs(), passwordModule.getPowerAuthPasswordJsModule());
    }

    public PowerAuthJsModule getPowerAuthJsModule() {
        return powerAuthJsModule;
    }

    // React integration

    @NonNull
    @Override
    public String getName() {
        return powerAuthJsModule.getName();
    }

    @ReactMethod
    public void isConfigured(@Nonnull String instanceId, final Promise promise) {
        powerAuthJsModule.isConfigured(instanceId, promise);
    }

    @ReactMethod
    public void configure(final String instanceId, final ReadableMap configuration, final ReadableMap clientConfiguration, final ReadableMap biometryConfiguration, final ReadableMap keychainConfiguration, final ReadableMap sharingConfiguration, Promise promise) {
        powerAuthJsModule.configure(instanceId, configuration, clientConfiguration, biometryConfiguration, keychainConfiguration, sharingConfiguration, promise);
    }

    @ReactMethod
    public void deconfigure(String instanceId, final Promise promise) {
        powerAuthJsModule.deconfigure(instanceId, promise);
    }

    @ReactMethod
    public void hasValidActivation(String instanceId, final Promise promise) {
        powerAuthJsModule.hasValidActivation(instanceId, promise);
    }

    @ReactMethod
    public void canStartActivation(String instanceId, final Promise promise) {
        powerAuthJsModule.canStartActivation(instanceId, promise);
    }

    @ReactMethod
    public void hasPendingActivation(String instanceId, final Promise promise) {
        powerAuthJsModule.hasPendingActivation(instanceId, promise);
    }

    @ReactMethod
    public void activationIdentifier(String instanceId, final Promise promise) {
        powerAuthJsModule.activationIdentifier(instanceId, promise);
    }

    @ReactMethod
    public  void activationFingerprint(String instanceId, final Promise promise) {
        powerAuthJsModule.activationFingerprint(instanceId, promise);
    }

    @ReactMethod
    public void getExternalPendingOperation(String instanceId, final Promise promise) {
        powerAuthJsModule.getExternalPendingOperation(instanceId, promise);
    }

    @ReactMethod
    public void fetchActivationStatus(String instanceId, final Promise promise) {
        powerAuthJsModule.fetchActivationStatus(instanceId, promise);
    }

    @ReactMethod
    public void createActivation(String instanceId, final ReadableMap activation, final Promise promise) {
        powerAuthJsModule.createActivation(instanceId, activation, promise);
    }

    @ReactMethod
    public void commitActivation(String instanceId, final ReadableMap authMap, final Promise promise) {
        powerAuthJsModule.commitActivation(instanceId, authMap, promise);
    }

    @ReactMethod
    public void removeActivationWithAuthentication(String instanceId,final ReadableMap authMap, final Promise promise) {
        powerAuthJsModule.removeActivationWithAuthentication(instanceId, authMap, promise);
    }

    @ReactMethod
    public void removeActivationLocal(String instanceId, final Promise promise) {
        powerAuthJsModule.removeActivationLocal(instanceId, promise);
    }

    @ReactMethod
    public void requestGetSignature(String instanceId, final ReadableMap authMap, final String uriId, @Nullable final ReadableMap params, final Promise promise) {
        powerAuthJsModule.requestGetSignature(instanceId, authMap, uriId, params, promise);
    }

    @ReactMethod
    public void requestSignature(String instanceId, final ReadableMap authMap, final String method, final String uriId, final  @Nullable String body, final Promise promise) {
        powerAuthJsModule.requestSignature(instanceId, authMap, method, uriId, body, promise);
    }

    @ReactMethod
    public void offlineSignature(String instanceId, final ReadableMap authMap, final String uriId, final  @Nullable String body, final String nonce, final Promise promise) {
        powerAuthJsModule.offlineSignature(instanceId, authMap, uriId, body, nonce, promise);
    }

    @ReactMethod
    public void verifyServerSignedData(String instanceId, final String data, final String signature, final boolean masterKey, final Promise promise) {
        powerAuthJsModule.verifyServerSignedData(instanceId, data, signature, masterKey, promise);
    }

    @ReactMethod
    public void unsafeChangePassword(String instanceId, final Dynamic oldPassword, final Dynamic newPassword, final Promise promise) {
        powerAuthJsModule.unsafeChangePassword(instanceId, oldPassword, newPassword, promise);
    }

    @ReactMethod
    public void changePassword(String instanceId, final Dynamic oldPassword, final Dynamic newPassword, final Promise promise) {
        powerAuthJsModule.changePassword(instanceId, oldPassword, newPassword, promise);
    }

    @ReactMethod
    public void addBiometryFactor(String instanceId, final Dynamic password, final ReadableMap prompt, final Promise promise) {
        powerAuthJsModule.addBiometryFactor(instanceId, password, prompt, promise);
    }

    @ReactMethod
    public void hasBiometryFactor(String instanceId, final Promise promise) {
        powerAuthJsModule.hasBiometryFactor(instanceId, promise);
    }

    @ReactMethod
    public void removeBiometryFactor(String instanceId, final Promise promise) {
        powerAuthJsModule.removeBiometryFactor(instanceId, promise);
    }

    @ReactMethod
    public void getBiometryInfo(String instanceId, final Promise promise) {
        powerAuthJsModule.getBiometryInfo(instanceId, promise);
    }

    @ReactMethod
    public void fetchEncryptionKey(String instanceId, final ReadableMap authMap, final int index, final Promise promise) {
        powerAuthJsModule.fetchEncryptionKey(instanceId, authMap, index, promise);
    }

    @ReactMethod
    public void signDataWithDevicePrivateKey(String instanceId, final ReadableMap authMap, final String data, final Promise promise) {
        powerAuthJsModule.signDataWithDevicePrivateKey(instanceId, authMap, data, promise);
    }

    @ReactMethod
    public void validatePassword(String instanceId, final Dynamic password, final Promise promise) {
        powerAuthJsModule.validatePassword(instanceId, password, promise);
    }

    @ReactMethod
    public void hasActivationRecoveryData(String instanceId, final Promise promise) {
        powerAuthJsModule.hasActivationRecoveryData(instanceId, promise);
    }

    @ReactMethod
    public void activationRecoveryData(String instanceId, final ReadableMap authMap, final Promise promise) {
        powerAuthJsModule.activationRecoveryData(instanceId, authMap, promise);
    }

    @ReactMethod
    public void confirmRecoveryCode(String instanceId, final String recoveryCode, final ReadableMap authMap, final Promise promise) {
        powerAuthJsModule.confirmRecoveryCode(instanceId, recoveryCode, authMap, promise);
    }

    @ReactMethod
    public void authenticateWithBiometry(String instanceId, final ReadableMap prompt, final boolean makeReusable, final Promise promise) {
        powerAuthJsModule.authenticateWithBiometry(instanceId, prompt, makeReusable, promise);
    }

    // TOKEN BASED AUTHENTICATION

    @ReactMethod
    public void requestAccessToken(String instanceId, final String tokenName, final ReadableMap authMap, final Promise promise) {
        powerAuthJsModule.requestAccessToken(instanceId, tokenName, authMap, promise);
    }

    @ReactMethod
    public void removeAccessToken(String instanceId, final String tokenName, final Promise promise) {
        powerAuthJsModule.removeAccessToken(instanceId, tokenName, promise);
    }

    @ReactMethod
    public void getLocalToken(String instanceId, final String tokenName, final Promise promise) {
        powerAuthJsModule.getLocalToken(instanceId, tokenName, promise);
    }

    @ReactMethod
    public void hasLocalToken(String instanceId, final String tokenName, final Promise promise) {
        powerAuthJsModule.hasLocalToken(instanceId, tokenName, promise);
    }

    @ReactMethod
    public void removeLocalToken(String instanceId, final String tokenName, final Promise promise) {
        powerAuthJsModule.removeLocalToken(instanceId, tokenName, promise);
    }

    @ReactMethod
    public void removeAllLocalTokens(String instanceId, final Promise promise) {
        powerAuthJsModule.removeAllLocalTokens(instanceId, promise);
    }

    @ReactMethod
    public void generateHeaderForToken(String instanceId, final String tokenName, final Promise promise) {
        powerAuthJsModule.generateHeaderForToken(instanceId, tokenName, promise);
    }

    // ACTIVATION CODE UTIL METHODS

    @ReactMethod
    public void parseActivationCode(String activationCode, final Promise promise) {
        powerAuthJsModule.parseActivationCode(activationCode, promise);
    }

    @ReactMethod
    public void validateActivationCode(String activationCode, final Promise promise) {
        powerAuthJsModule.validateActivationCode(activationCode, promise);
    }

    @ReactMethod
    public void parseRecoveryCode(String recoveryCode, final Promise promise) {
        powerAuthJsModule.parseRecoveryCode(recoveryCode, promise);
    }

    @ReactMethod
    public void validateRecoveryCode(String recoveryCode, final Promise promise) {
        powerAuthJsModule.validateRecoveryCode(recoveryCode, promise);
    }

    @ReactMethod
    public void validateRecoveryPuk(String puk, final Promise promise) {
        powerAuthJsModule.validateRecoveryPuk(puk, promise);
    }

    @ReactMethod
    public void validateTypedCharacter(int character, final Promise promise) {
        powerAuthJsModule.validateTypedCharacter(character, promise);
    }

    @ReactMethod
    public void correctTypedCharacter(int character, final Promise promise) {
        powerAuthJsModule.correctTypedCharacter(character, promise);
    }
}

File: other-platforms-support/cordova/android/src/main/java/com/wultra/android/powerauth/cdv/PowerAuthModule.kt

package com.wultra.android.powerauth.cordova.plugin

import android.app.Activity
import android.util.Log
import com.wultra.android.powerauth.bridge.getDynamic
import com.wultra.android.powerauth.bridge.getOptReadableMap
import com.wultra.android.powerauth.bridge.getOptString
import com.wultra.android.powerauth.bridge.getReadableMap
import com.wultra.android.powerauth.cdv.util.Promise
import com.wultra.android.powerauth.js.ActivityProvider
import com.wultra.android.powerauth.js.PowerAuthJsModule
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaInterface
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.CordovaWebView
import org.json.JSONArray
import org.json.JSONException

class PowerAuthModule : CordovaPlugin() {

    companion object {
        const val TAG = "PowerAuthModule"
    }

    private lateinit var powerAuthJsModule: PowerAuthJsModule

    private val activityProvider: ActivityProvider = object : ActivityProvider {
        override fun getActivity(): Activity = cordova.activity
    }

    override fun initialize(cordova: CordovaInterface, webView: CordovaWebView) {
        super.initialize(cordova, webView);
        val powerAuthObjectRegister = webView.pluginManager.getPlugin("PowerAuthObjectRegister") as PowerAuthObjectRegister
        val powerAuthPasswordModule = webView.pluginManager.getPlugin("PowerAuthPasswordModule") as PowerAuthPasswordModule
        powerAuthJsModule = PowerAuthJsModule(cordova.activity, activityProvider, powerAuthObjectRegister.objectRegisterJs, powerAuthPasswordModule.powerAuthPasswordJsModule)
    }

    @Throws(JSONException::class)
    override fun execute(action: String, args: JSONArray, callbackContext: CallbackContext): Boolean {
        val promise = Promise(callbackContext)
        when (action) {
            "isConfigured" -> {
                isConfigured(args, promise)
                return true;
            }
            "configure" -> {
                configure(args, promise)
                return true;
            }
            "deconfigure" -> {
                deconfigure(args, promise)
                return true;
            }
            "hasValidActivation" -> {
                hasValidActivation(args, promise)
                return true;
            }
            "canStartActivation" -> {
                canStartActivation(args, promise)
                return true;
            }
            "hasPendingActivation" -> {
                hasPendingActivation(args, promise)
                return true;
            }
            "activationIdentifier" -> {
                activationIdentifier(args, promise)
                return true;
            }
            "activationFingerprint" -> {
                activationFingerprint(args, promise)
                return true;
            }
            "getExternalPendingOperation" -> {
                getExternalPendingOperation(args, promise)
                return true;
            }
            "fetchActivationStatus" -> {
                fetchActivationStatus(args, promise)
                return true;
            }
            "createActivation" -> {
                createActivation(args, promise)
                return true;
            }
            "commitActivation" -> {
                commitActivation(args, promise)
                return true;
            }
            "removeActivationWithAuthentication" -> {
                removeActivationWithAuthentication(args, promise)
                return true;
            }
            "removeActivationLocal" -> {
                removeActivationLocal(args, promise)
                return true;
            }
            "requestGetSignature" -> {
                requestGetSignature(args, promise)
                return true;
            }
            "requestSignature" -> {
                requestSignature(args, promise)
                return true;
            }
            "offlineSignature" -> {
                offlineSignature(args, promise)
                return true;
            }
            "verifyServerSignedData" -> {
                verifyServerSignedData(args, promise)
                return true;
            }
            "unsafeChangePassword" -> {
                unsafeChangePassword(args, promise)
                return true;
            }
            "changePassword" -> {
                changePassword(args, promise)
                return true;
            }
            "addBiometryFactor" -> {
                addBiometryFactor(args, promise)
                return true;
            }
            "hasBiometryFactor" -> {
                hasBiometryFactor(args, promise)
                return true;
            }
            "removeBiometryFactor" -> {
                removeBiometryFactor(args, promise)
                return true;
            }
            "getBiometryInfo" -> {
                getBiometryInfo(args, promise)
                return true;
            }
            "fetchEncryptionKey" -> {
                fetchEncryptionKey(args, promise)
                return true;
            }
            "signDataWithDevicePrivateKey" -> {
                signDataWithDevicePrivateKey(args, promise)
                return true;
            }
            "validatePassword" -> {
                validatePassword(args, promise)
                return true;
            }
            "hasActivationRecoveryData" -> {
                hasActivationRecoveryData(args, promise)
                return true;
            }
            "activationRecoveryData" -> {
                activationRecoveryData(args, promise)
                return true;
            }
            "confirmRecoveryCode" -> {
                confirmRecoveryCode(args, promise)
                return true;
            }
            "authenticateWithBiometry" -> {
                authenticateWithBiometry(args, promise)
                return true;
            }
            // TOKEN BASED AUTHENTICATION
            "requestAccessToken" -> {
                requestAccessToken(args, promise)
                return true;
            }
            "removeAccessToken" -> {
                removeAccessToken(args, promise)
                return true;
            }
            "getLocalToken" -> {
                getLocalToken(args, promise)
                return true;
            }
            "hasLocalToken" -> {
                hasLocalToken(args, promise)
                return true;
            }
            "removeLocalToken" -> {
                removeLocalToken(args, promise)
                return true;
            }
            "removeAllLocalTokens" -> {
                removeAllLocalTokens(args, promise)
                return true;
            }
            "generateHeaderForToken" -> {
                generateHeaderForToken(args, promise)
                return true;
            }
            // ACTIVATION CODE UTIL METHODS
            "parseActivationCode" -> {
                parseActivationCode(args, promise)
                return true;
            }
            "validateActivationCode" -> {
                validateActivationCode(args, promise)
                return true;
            }
            "parseRecoveryCode" -> {
                parseRecoveryCode(args, promise)
                return true;
            }
            "validateRecoveryCode" -> {
                validateRecoveryCode(args, promise)
                return true;
            }
            "validateRecoveryPuk" -> {
                validateRecoveryPuk(args, promise)
                return true;
            }
            "validateTypedCharacter" -> {
                validateTypedCharacter(args, promise)
                return true;
            }
            "correctTypedCharacter" -> {
                correctTypedCharacter(args, promise)
                return true;
            }
        }
        return false  // Returning false results in a "MethodNotFound" error.
    }

    private fun isConfigured(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.isConfigured(instanceId, promise);
    }

    private fun configure(args: JSONArray, promise: Promise) {
        Log.d(TAG, "configure(args=$args)")
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            val configuration = args.getReadableMap(1)
            val clientConfiguration = args.getReadableMap(2)
            val biometryConfiguration = args.getReadableMap(3)
            val keychainConfiguration = args.getReadableMap(4)
            val sharingConfiguration = args.getReadableMap(5)
            powerAuthJsModule.configure(instanceId, configuration, clientConfiguration, biometryConfiguration, keychainConfiguration, sharingConfiguration, promise);
        })
    }

    private fun deconfigure(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.deconfigure(instanceId, promise);
    }

    private fun hasValidActivation(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.hasValidActivation(instanceId, promise);
    }

    private fun canStartActivation(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.canStartActivation(instanceId, promise);
    }

    private fun hasPendingActivation(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.hasPendingActivation(instanceId, promise);
    }

    private fun activationIdentifier(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.activationIdentifier(instanceId, promise);
    }

    private fun activationFingerprint(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.activationFingerprint(instanceId, promise);
    }

    private fun getExternalPendingOperation(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.getExternalPendingOperation(instanceId, promise);
    }

    private fun fetchActivationStatus(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.fetchActivationStatus(instanceId, promise);
    }

    private fun createActivation(args: JSONArray, promise: Promise) {
        Log.d(TAG, "createActivation(args=$args)")
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            val activation = args.getReadableMap(1)
            powerAuthJsModule.createActivation(instanceId, activation, promise);
        })
    }

    private fun commitActivation(args: JSONArray, promise: Promise) {
        Log.d(TAG, "commitActivation(args=$args)")
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        powerAuthJsModule.commitActivation(instanceId, authMap, promise);
    }

    private fun removeActivationWithAuthentication(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        powerAuthJsModule.removeActivationWithAuthentication(instanceId, authMap, promise);
    }

    private fun removeActivationLocal(args: JSONArray, promise: Promise) {
        Log.d(TAG, "removeActivationLocal(args=$args)")
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            powerAuthJsModule.removeActivationLocal(instanceId, promise);
        })
    }

    private fun requestGetSignature(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        val uriId = args.getOptString(2)
        val params = args.getOptReadableMap(3)
        powerAuthJsModule.requestGetSignature(instanceId, authMap, uriId, params, promise);
    }

    private fun requestSignature(args: JSONArray, promise: Promise) {
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            val authMap = args.getReadableMap(1)
            val method = args.getOptString(2)
            val uriId = args.getOptString(3)
            val body = args.getOptString(4)
            powerAuthJsModule.requestSignature(instanceId, authMap, method, uriId, body, promise);
        })
    }

    private fun offlineSignature(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        val uriId = args.getString(2)
        val body = args.getOptString(3)
        val nonce = args.getString(4)
        powerAuthJsModule.offlineSignature(instanceId, authMap, uriId, body, nonce, promise);
    }

    private fun verifyServerSignedData(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val `data` = args.getString(1)
        val signature = args.getOptString(2)
        val masterKey = args.getBoolean(3)
        powerAuthJsModule.verifyServerSignedData(instanceId, data, signature, masterKey, promise);
    }

    private fun unsafeChangePassword(args: JSONArray, promise: Promise) {
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            val oldPassword = args.getDynamic(1)
            val newPassword = args.getDynamic(2)
            powerAuthJsModule.unsafeChangePassword(instanceId, oldPassword, newPassword, promise);
        })
    }

    private fun changePassword(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val oldPassword = args.getDynamic(1)
        val newPassword = args.getDynamic(2)
        powerAuthJsModule.changePassword(instanceId, oldPassword, newPassword, promise);
    }

    private fun addBiometryFactor(args: JSONArray, promise: Promise) {
        Log.i(TAG, "addBiometryFactor(ags=$args)")
        val instanceId = args.getString(0)
        val password = args.getDynamic(1)
        val prompt = args.getOptReadableMap(2)
        powerAuthJsModule.addBiometryFactor(instanceId, password, prompt, promise);
    }

    private fun hasBiometryFactor(args: JSONArray, promise: Promise) {
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            powerAuthJsModule.hasBiometryFactor(instanceId, promise);
        })
    }

    private fun removeBiometryFactor(args: JSONArray, promise: Promise) {
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            powerAuthJsModule.removeBiometryFactor(instanceId, promise);
        })
    }

    private fun getBiometryInfo(args: JSONArray, promise: Promise) {
        val instanceId = args.getOptString(0)
        powerAuthJsModule.getBiometryInfo(instanceId, promise);
    }

    private fun fetchEncryptionKey(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        val index = args.getInt(2)
        powerAuthJsModule.fetchEncryptionKey(instanceId, authMap, index, promise);
    }

    private fun signDataWithDevicePrivateKey(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        val `data` = args.getString(2)
        powerAuthJsModule.signDataWithDevicePrivateKey(instanceId, authMap, `data`, promise);
    }

    private fun validatePassword(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val password = args.getDynamic(1)
        powerAuthJsModule.validatePassword(instanceId, password, promise);
    }

    private fun hasActivationRecoveryData(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.hasActivationRecoveryData(instanceId, promise);
    }

    private fun activationRecoveryData(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val authMap = args.getReadableMap(1)
        powerAuthJsModule.activationRecoveryData(instanceId, authMap, promise);
    }

    private fun confirmRecoveryCode(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val recoveryCode = args.getString(1)
        val authMap = args.getReadableMap(2)
        powerAuthJsModule.confirmRecoveryCode(instanceId, recoveryCode, authMap, promise);
    }

    private fun authenticateWithBiometry(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val prompt = args.getOptReadableMap(1)
        val makeReusable = args.getBoolean(2)
        powerAuthJsModule.authenticateWithBiometry(instanceId, prompt, makeReusable, promise);
    }

    // TOKEN BASED AUTHENTICATION

    private fun requestAccessToken(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val tokenName = args.getString(1)
        val authMap = args.getReadableMap(2)
        powerAuthJsModule.requestAccessToken(instanceId, tokenName, authMap, promise);
    }

    private fun removeAccessToken(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val tokenName = args.getString(1)
        powerAuthJsModule.removeAccessToken(instanceId, tokenName, promise);
    }

    private fun getLocalToken(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val tokenName = args.getString(1)
        powerAuthJsModule.getLocalToken(instanceId, tokenName, promise);
    }

    private fun hasLocalToken(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val tokenName = args.getString(1)
        powerAuthJsModule.hasLocalToken(instanceId, tokenName, promise);
    }

    private fun removeLocalToken(args: JSONArray, promise: Promise) {
        cordova.threadPool.execute(Runnable {
            val instanceId = args.getString(0)
            val tokenName = args.getString(1)
            powerAuthJsModule.removeLocalToken(instanceId, tokenName, promise);
        })
    }

    private fun removeAllLocalTokens(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        powerAuthJsModule.removeAllLocalTokens(instanceId, promise);
    }

    private fun generateHeaderForToken(args: JSONArray, promise: Promise) {
        val instanceId = args.getString(0)
        val tokenName = args.getString(1)
        powerAuthJsModule.generateHeaderForToken(instanceId, tokenName, promise);
    }

    // ACTIVATION CODE UTIL METHODS

    private fun parseActivationCode(args: JSONArray, promise: Promise) {
        val activationCode = args.getString(0)
        powerAuthJsModule.parseActivationCode(activationCode, promise);
    }

    private fun validateActivationCode(args: JSONArray, promise: Promise) {
        val activationCode = args.getString(0)
        powerAuthJsModule.validateActivationCode(activationCode, promise);
    }

    private fun parseRecoveryCode(args: JSONArray, promise: Promise) {
        val recoveryCode = args.getString(0)
        powerAuthJsModule.parseRecoveryCode(recoveryCode, promise);
    }

    private fun validateRecoveryCode(args: JSONArray, promise: Promise) {
        val recoveryCode = args.getString(0)
        powerAuthJsModule.validateRecoveryCode(recoveryCode, promise);
    }

    private fun validateRecoveryPuk(args: JSONArray, promise: Promise) {
        val puk = args.getString(0)
        powerAuthJsModule.validateRecoveryPuk(puk, promise);
    }

    private fun validateTypedCharacter(args: JSONArray, promise: Promise) {
        val character = args.getInt(0)
        powerAuthJsModule.validateTypedCharacter(character, promise);
    }

    private fun correctTypedCharacter(args: JSONArray, promise: Promise) {
        val character = args.getInt(0)
        powerAuthJsModule.correctTypedCharacter(character, promise);
    }
}